File: android/src/main/java/com/noesis/diagnostic/modules/LocationModule.java
package com.noesis.diagnostic.modules;

import android.Manifest;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.pm.PackageManager;
import android.hardware.Sensor;
import android.hardware.SensorManager;
import android.location.LocationManager;
import android.net.Uri;
import android.os.Build;
import android.provider.Settings;
import android.util.Log;

import androidx.core.content.ContextCompat;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;

public class LocationModule {

    private static final String PREFS = "DiagCapPrefs";
    private static final String KEY_LOC_ASKED = "loc_asked";

    private final Plugin plugin;

    public LocationModule(Plugin plugin) {
        this.plugin = plugin;
    }

    private Context getContext() {
        return plugin.getContext();
    }

    /*
     * Tracks whether the user has ever been asked for location.
     * We persist this ourselves because Android doesn't give us a clean
     * "never asked" vs "denied" distinction — that gap only shows up via
     * shouldShowRequestPermissionRationale(), which is unreliable as a first-read.
     */
    private boolean wasLocationEverAsked() {
        SharedPreferences sp = getContext().getSharedPreferences(PREFS, Context.MODE_PRIVATE);
        return sp.getBoolean(KEY_LOC_ASKED, false);
    }

    public void setLocationEverAsked() {
        SharedPreferences sp = getContext().getSharedPreferences(PREFS, Context.MODE_PRIVATE);
        sp.edit().putBoolean(KEY_LOC_ASKED, true).apply();
    }

    /*
     * Returns the full authorization status string matching Cordova:
     * "authorized_always", "authorized_when_in_use", "denied", "not_determined".
     *
     * Background location is only tracked on Android Q+. On older versions,
     * foreground grant is the highest state we can reach.
     */
    public void getLocationAuthorizationStatus(PluginCall call) {
        boolean fineGranted =
            ContextCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED;

        boolean coarseGranted =
            ContextCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED;

        boolean foregroundGranted = fineGranted || coarseGranted;

        boolean backgroundGranted = false;
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            backgroundGranted =
                ContextCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_BACKGROUND_LOCATION) == PackageManager.PERMISSION_GRANTED;
        }

        JSObject ret = new JSObject();

        if (backgroundGranted) {
            ret.put("status", "authorized_always");
        } else if (foregroundGranted) {
            ret.put("status", "authorized_when_in_use");
        } else {
            ret.put("status", wasLocationEverAsked() ? "denied" : "not_determined");
        }

        Log.d("DiagCap", "getLocationAuthorizationStatus -> " + ret.getString("status"));
        call.resolve(ret);
    }

    /*
     * Returns { available: boolean }.
     * Location is "available" only if both the system provider is on AND
     * the app has at least foreground permission. Matches Cordova behavior.
     */
    public void isLocationAvailable(PluginCall call) {
        boolean available = false;

        try {
            int mode = getLocationModeInt();
            boolean enabled = (mode != 0);
            boolean authorized = isLocationAuthorizedForeground();
            available = enabled && authorized;
        } catch (Exception ignored) {}

        JSObject ret = new JSObject();
        ret.put("available", available);
        Log.d("DiagCap", "isLocationAvailable -> " + available);
        call.resolve(ret);
    }

    /*
     * Returns { enabled: boolean }.
     * Checks only the system location provider — does not factor in app permission.
     */
    public void isLocationEnabled(PluginCall call) {
        boolean enabled = false;

        try {
            int mode = getLocationModeInt();
            enabled = (mode != 0);
        } catch (Exception ignored) {}

        JSObject ret = new JSObject();
        ret.put("enabled", enabled);
        Log.d("DiagCap", "isLocationEnabled -> " + enabled);
        call.resolve(ret);
    }

    /*
     * Opens the app's detail settings page (not the global location settings).
     * This lets the user grant/revoke location permission for this specific app.
     */
    public void openLocationSettings(PluginCall call) {
        Intent intent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
        Uri uri = Uri.fromParts("package", getContext().getPackageName(), null);
        intent.setData(uri);

        if (plugin.getActivity() != null) {
            plugin.getActivity().startActivity(intent);
        } else {
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            getContext().startActivity(intent);
        }

        call.resolve();
    }

    /*
     * Returns { mode: string } — one of: "high_accuracy", "device_only", "battery_saving", "location_off", "unknown".
     *
     * Pre-API 28 reads Settings.Secure.LOCATION_MODE directly.
     * API 28+ that setting was deprecated — we reconstruct the mode by
     * querying GPS and network providers from LocationManager.
     */
    public void getLocationMode(PluginCall call) {
        String modeName = "unknown";

        try {
            int mode = getLocationModeInt();

            modeName = switch (mode) {
                case 3 -> "high_accuracy";
                case 1 -> "device_only";
                case 2 -> "battery_saving";
                case 0 -> "location_off";
                default -> "unknown";
            };
        } catch (Exception ignored) {}

        JSObject ret = new JSObject();
        ret.put("mode", modeName);
        Log.d("DiagCap", "getLocationMode -> " + modeName);
        call.resolve(ret);
    }

    /*
     * Returns { enabled: boolean } — true if GPS provider is active.
     * Mode 3 (high_accuracy) and mode 1 (device_only) both have GPS on.
     */
    public void isGpsLocationEnabled(PluginCall call) {
        boolean enabled = false;

        try {
            int mode = getLocationModeInt();
            enabled = (mode == 3 || mode == 1);
        } catch (Exception ignored) {}

        JSObject ret = new JSObject();
        ret.put("enabled", enabled);
        Log.d("DiagCap", "isGpsLocationEnabled -> " + enabled);
        call.resolve(ret);
    }

    /*
     * Returns { enabled: boolean } — true if network/WiFi location provider is active.
     * Mode 3 (high_accuracy) and mode 2 (battery_saving) both have network on.
     */
    public void isNetworkLocationEnabled(PluginCall call) {
        boolean enabled = false;

        try {
            int mode = getLocationModeInt();
            enabled = (mode == 3 || mode == 2);
        } catch (Exception ignored) {}

        JSObject ret = new JSObject();
        ret.put("enabled", enabled);
        Log.d("DiagCap", "isNetworkLocationEnabled -> " + enabled);
        call.resolve(ret);
    }

    /*
     * Returns { available: boolean } — GPS enabled AND app has foreground permission.
     */
    public void isGpsLocationAvailable(PluginCall call) {
        boolean available = false;

        try {
            int mode = getLocationModeInt();
            boolean gpsEnabled = (mode == 3 || mode == 1);
            boolean authorized = isLocationAuthorizedForeground();
            available = gpsEnabled && authorized;
        } catch (Exception ignored) {}

        JSObject ret = new JSObject();
        ret.put("available", available);
        Log.d("DiagCap", "isGpsLocationAvailable -> " + available);
        call.resolve(ret);
    }

    /*
     * Returns { available: boolean } — network location enabled AND app has foreground permission.
     */
    public void isNetworkLocationAvailable(PluginCall call) {
        boolean available = false;

        try {
            int mode = getLocationModeInt();
            boolean netEnabled = (mode == 3 || mode == 2);
            boolean authorized = isLocationAuthorizedForeground();
            available = netEnabled && authorized;
        } catch (Exception ignored) {}

        JSObject ret = new JSObject();
        ret.put("available", available);
        Log.d("DiagCap", "isNetworkLocationAvailable -> " + available);
        call.resolve(ret);
    }

    /*
     * Opens the system's location source settings screen.
     * Unlike openLocationSettings, this is the global toggle — not app-specific.
     */
    public void switchToLocationSettings(PluginCall call) {
        Intent intent = new Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS);

        if (plugin.getActivity() != null) {
            plugin.getActivity().startActivity(intent);
        } else {
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
            getContext().startActivity(intent);
        }

        Log.d("DiagCap", "switchToLocationSettings -> opened");
        call.resolve();
    }

    /*
     * Returns { available: boolean } — checks for a magnetic field sensor (hardware compass).
     * Purely hardware capability check — no permission involved.
     */
    public void isCompassAvailable(PluginCall call) {
        boolean available = false;

        try {
            SensorManager sm = (SensorManager) getContext().getSystemService(Context.SENSOR_SERVICE);
            available = (sm != null && sm.getDefaultSensor(Sensor.TYPE_MAGNETIC_FIELD) != null);
        } catch (Exception ignored) {}

        JSObject ret = new JSObject();
        ret.put("available", available);
        Log.d("DiagCap", "isCompassAvailable -> " + available);
        call.resolve(ret);
    }

    /*
     * Returns { value: boolean } — true if the app has either fine or coarse location granted.
     * "Authorized" here means foreground only — background is not checked.
     */
    public void isLocationAuthorized(PluginCall call) {
        boolean authorized = isLocationAuthorizedForeground();
        JSObject ret = new JSObject();
        ret.put("value", authorized);
        Log.d("DiagCap", "isLocationAuthorized -> " + authorized);
        call.resolve(ret);
    }

    /*
     * Returns { value: "full" } always on Android.
     * The reduced accuracy concept (iOS 14+) doesn't exist on Android —
     * this is here purely for Cordova API surface parity.
     */
    public void getLocationAccuracyAuthorization(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("value", "full");
        Log.d("DiagCap", "getLocationAccuracyAuthorization -> full");
        call.resolve(ret);
    }

    /*
     * Returns { value: "full" } always on Android.
     * Temporary accuracy downgrade/upgrade is an iOS 14+ concept.
     * Android always runs at full accuracy if permission is granted.
     */
    public void requestTemporaryFullAccuracyAuthorization(PluginCall call) {
        JSObject ret = new JSObject();
        ret.put("value", "full");
        Log.d("DiagCap", "requestTemporaryFullAccuracyAuthorization -> full");
        call.resolve(ret);
    }

    /*
     * Called after the Capacitor permission dialog closes for foreground location.
     * On Android Q (10), we have to fire the background location prompt separately
     * if "always" was requested — Android Q requires a second prompt for background.
     * Post-Q, "always" requires the user to manually go to Settings.
     */
    public void onLocationPermissionResult(PluginCall call) {
        boolean fineGranted =
            ContextCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED;

        boolean coarseGranted =
            ContextCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED;

        boolean foregroundGranted = fineGranted || coarseGranted;

        if (!foregroundGranted) {
            JSObject ret = new JSObject();
            ret.put("status", "denied");
            Log.d("DiagCap", "onLocationPermissionResult -> " + ret.getString("status"));
            call.resolve(ret);
            return;
        }

        // If background already granted, keep authorized_always regardless of requested mode
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            boolean backgroundGranted =
                ContextCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_BACKGROUND_LOCATION)
                    == PackageManager.PERMISSION_GRANTED;

            if (backgroundGranted) {
                JSObject ret = new JSObject();
                ret.put("status", "authorized_always");
                Log.d("DiagCap", "onLocationPermissionResult -> authorized_always (already had background)");
                call.resolve(ret);
                return;
            }
        }

        String mode = call.getData().optString("mode", "when_in_use");

        if ("always".equalsIgnoreCase(mode) && Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {

            if (Build.VERSION.SDK_INT == Build.VERSION_CODES.Q) {
                JSObject ret = new JSObject();
                ret.put("status", "authorized_when_in_use");
                Log.d("DiagCap", "onLocationPermissionResult -> " + ret.getString("status") + " (background prompt requested next)");
                call.resolve(ret);
                return;
            }

            JSObject ret = new JSObject();
            ret.put("status", "authorized_when_in_use");
            Log.d("DiagCap", "onLocationPermissionResult -> " + ret.getString("status") + " (settings required for always)");
            call.resolve(ret);
            return;
        }

        JSObject ret = new JSObject();
        ret.put("status", "authorized_when_in_use");
        Log.d("DiagCap", "onLocationPermissionResult -> " + ret.getString("status"));
        call.resolve(ret);
    }

    /*
     * Called after the background location permission prompt closes (Android Q only path).
     * Returns "authorized_always" if background was granted, "authorized_when_in_use" otherwise.
     */
    public void onBackgroundLocationPermissionResult(PluginCall call) {
        boolean backgroundGranted = false;

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
            backgroundGranted =
                ContextCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_BACKGROUND_LOCATION) == PackageManager.PERMISSION_GRANTED;
        }

        JSObject ret = new JSObject();
        ret.put("status", backgroundGranted ? "authorized_always" : "authorized_when_in_use");
        Log.d("DiagCap", "onBackgroundLocationPermissionResult -> " + ret.getString("status"));
        call.resolve(ret);
    }

    // -------------------------------------------------------------------------
    // Internal helpers
    // -------------------------------------------------------------------------

    /*
     * Returns an integer representing the active location mode.
     * Maps to: 0=off, 1=device_only (GPS), 2=battery_saving (network), 3=high_accuracy (GPS+network).
     *
     * Pre-API 28: reads Settings.Secure.LOCATION_MODE directly (deprecated but still works).
     * API 28+: reconstructs mode by querying GPS and network providers from LocationManager.
     */
    private int getLocationModeInt() throws Exception {
        if (Build.VERSION.SDK_INT < 28) {
            return Settings.Secure.getInt(
                getContext().getContentResolver(),
                Settings.Secure.LOCATION_MODE
            );
        }

        LocationManager lm = (LocationManager) getContext().getSystemService(Context.LOCATION_SERVICE);
        boolean gps = false;
        boolean network = false;

        if (lm != null) {
            try { gps = lm.isProviderEnabled(LocationManager.GPS_PROVIDER); } catch (Exception ignored) {}
            try { network = lm.isProviderEnabled(LocationManager.NETWORK_PROVIDER); } catch (Exception ignored) {}
        }

        if (gps && network) return 3;
        if (gps) return 1;
        if (network) return 2;
        return 0;
    }

    private boolean isLocationAuthorizedForeground() {
        boolean fineGranted =
            ContextCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_FINE_LOCATION) == PackageManager.PERMISSION_GRANTED;
        boolean coarseGranted =
            ContextCompat.checkSelfPermission(getContext(), Manifest.permission.ACCESS_COARSE_LOCATION) == PackageManager.PERMISSION_GRANTED;

        return fineGranted || coarseGranted;
    }
}

File: ios/Sources/DiagnosticPlugin/modules/LocationModule.swift
import Foundation
import Capacitor
import CoreLocation
import UIKit

final class LocationModule: NSObject, CLLocationManagerDelegate {

    private let location_manager = CLLocationManager()

    // Holds the pending call from requestLocationAuthorization() until the
    // delegate fires a status change. Set to nil once resolved.
    private var pending_location_auth_call: CAPPluginCall?
    private var current_location_authorization_status: String?

    // iOS 14+ only — tracks accuracy authorization separately from location permission
    @available(iOS 14.0, *)
    private var current_location_accuracy_authorization: String?

    override init() {
        super.init()
        location_manager.delegate = self
    }

    // -------------------------------------------------------------------------
    // Helpers
    // -------------------------------------------------------------------------

    /*
     * iOS 14+ uses instance property authorizationStatus.
     * Before 14, the class method is the only option.
     */
    private func get_authorization_status() -> CLAuthorizationStatus {
        if #available(iOS 14.0, *) {
            return location_manager.authorizationStatus
        } else {
            return CLLocationManager.authorizationStatus()
        }
    }

    /*
     * Maps CLAuthorizationStatus to Cordova-compatible strings.
     * Note: on iOS there's no "not_determined" for foreground/background split —
     * "authorizedAlways" maps to "granted" to match Cordova's iOS behavior.
     */
    private func auth_status_string(_ auth_status: CLAuthorizationStatus) -> String {
        switch auth_status {
        case .denied, .restricted:
            return "denied"
        case .notDetermined:
            return "not_determined"
        case .authorizedAlways:
            return "granted"
        case .authorizedWhenInUse:
            return "authorized_when_in_use"
        @unknown default:
            return "not_determined"
        }
    }

    private func is_location_authorized_internal() -> Bool {
        let status = auth_status_string(get_authorization_status())
        return status == "granted" || status == "authorized_when_in_use"
    }

    private func open_app_settings() {
        guard let url = URL(string: UIApplication.openSettingsURLString) else { return }
        DispatchQueue.main.async {
            UIApplication.shared.open(url, options: [:], completionHandler: nil)
        }
    }

    /*
     * Called by the CLLocationManagerDelegate whenever auth status changes.
     * Resolves the pending call if one exists, but only if the status actually changed.
     * This mirrors Cordova's behavior of returning only on state change.
     */
    private func report_change_authorization_status(_ auth_status: CLAuthorizationStatus) {
        let status = auth_status_string(auth_status)

        let changed: Bool
        if let current = current_location_authorization_status {
            changed = (status != current)
        } else {
            changed = true
        }
        current_location_authorization_status = status

        if changed {
            if let call = pending_location_auth_call {
                pending_location_auth_call = nil
                call.resolve(["status": status])
            }
        }
    }

    // -------------------------------------------------------------------------
    // Plugin methods
    // -------------------------------------------------------------------------

    /*
     * Returns { status: string } — current authorization status without prompting.
     */
    func getLocationAuthorizationStatus(_ call: CAPPluginCall) {
        call.resolve(["status": auth_status_string(get_authorization_status())])
    }

    /*
     * Triggers the location permission prompt if status is "not_determined".
     * If already determined (any status), resolves immediately with current status —
     * this avoids the call hanging when the user doesn't interact with the dialog,
     * which is a known Cordova quirk we preserve here.
     *
     * Requires NSLocationWhenInUseUsageDescription in Info.plist.
     * "always" mode additionally requires NSLocationAlwaysAndWhenInUseUsageDescription.
     */
    func requestLocationAuthorization(_ call: CAPPluginCall) {
        let mode = call.getString("mode") ?? "when_in_use"
        let always = (mode == "always")

        pending_location_auth_call = call

        let current = auth_status_string(get_authorization_status())
        if current != "not_determined" {
            pending_location_auth_call = nil
            call.resolve(["status": current])
            return
        }

        if always {
            let has_key = Bundle.main.object(forInfoDictionaryKey: "NSLocationAlwaysAndWhenInUseUsageDescription") != nil
            if !has_key {
                pending_location_auth_call = nil
                call.reject("Missing NSLocationAlwaysAndWhenInUseUsageDescription in Info.plist")
                return
            }
            location_manager.requestAlwaysAuthorization()
        } else {
            let has_key = Bundle.main.object(forInfoDictionaryKey: "NSLocationWhenInUseUsageDescription") != nil
            if !has_key {
                pending_location_auth_call = nil
                call.reject("Missing NSLocationWhenInUseUsageDescription in Info.plist")
                return
            }
            location_manager.requestWhenInUseAuthorization()
        }
    }

    /*
     * Returns { enabled: boolean } — whether location services are enabled at the system level.
     * This is the iOS equivalent of Android's "location mode on/off" check.
     * Does not check app-level permission.
     */
    func isLocationEnabled(_ call: CAPPluginCall) {
        call.resolve(["enabled": CLLocationManager.locationServicesEnabled()])
    }

    /*
     * Returns { available: boolean } — location services are on AND the app is authorized.
     */
    func isLocationAvailable(_ call: CAPPluginCall) {
        let value = CLLocationManager.locationServicesEnabled() && is_location_authorized_internal()
        call.resolve(["available": value])
    }

    /*
     * Returns { mode: string }.
     * iOS doesn't have distinct location modes like Android (GPS only, network only, etc.).
     * If services are enabled we return "high_accuracy" to match Cordova's iOS output.
     */
    func getLocationMode(_ call: CAPPluginCall) {
        if !CLLocationManager.locationServicesEnabled() {
            call.resolve(["mode": "location_off"])
        } else {
            call.resolve(["mode": "high_accuracy"])
        }
    }

    /*
     * iOS doesn't separate GPS and network providers — both map to CLLocationManager.
     * These three methods all check the same underlying locationServicesEnabled() flag
     * to maintain Cordova API parity across platforms.
     */
    func isGpsLocationEnabled(_ call: CAPPluginCall) {
        call.resolve(["enabled": CLLocationManager.locationServicesEnabled()])
    }

    func isNetworkLocationEnabled(_ call: CAPPluginCall) {
        call.resolve(["enabled": CLLocationManager.locationServicesEnabled()])
    }

    func isGpsLocationAvailable(_ call: CAPPluginCall) {
        let value = CLLocationManager.locationServicesEnabled() && is_location_authorized_internal()
        call.resolve(["available": value])
    }

    func isNetworkLocationAvailable(_ call: CAPPluginCall) {
        let value = CLLocationManager.locationServicesEnabled() && is_location_authorized_internal()
        call.resolve(["available": value])
    }

    /*
     * Opens the app's page in Settings. Both open/switch to location settings
     * go to the same place on iOS — no direct link to the global location toggle.
     */
    func openLocationSettings(_ call: CAPPluginCall) {
        open_app_settings()
        call.resolve()
    }

    func switchToLocationSettings(_ call: CAPPluginCall) {
        open_app_settings()
        call.resolve()
    }

    /*
     * Returns { available: boolean } — checks for hardware compass (heading availability).
     */
    func isCompassAvailable(_ call: CAPPluginCall) {
        call.resolve(["available": CLLocationManager.headingAvailable()])
    }

    /*
     * Returns { value: boolean } — true if app has either whenInUse or always authorization.
     */
    func isLocationAuthorized(_ call: CAPPluginCall) {
        call.resolve(["value": is_location_authorized_internal()])
    }

    /*
     * Returns { value: "full" | "reduced" }.
     * iOS 14+ introduced reduced accuracy ("approximate location"). Below iOS 14 always returns "full".
     */
    func getLocationAccuracyAuthorization(_ call: CAPPluginCall) {
        if #available(iOS 14.0, *) {
            let accuracy = (location_manager.accuracyAuthorization == .fullAccuracy) ? "full" : "reduced"
            call.resolve(["value": accuracy])
        } else {
            call.resolve(["value": "full"])
        }
    }

    /*
     * Requests a temporary upgrade to full accuracy using a purpose key from Info.plist.
     * Only available on iOS 14+. Rejects cleanly on older versions.
     * Requires NSLocationTemporaryUsageDescriptionDictionary in Info.plist with a matching key.
     *
     * @param purpose — string key matching an entry in NSLocationTemporaryUsageDescriptionDictionary
     */
    func requestTemporaryFullAccuracyAuthorization(_ call: CAPPluginCall) {
        if #available(iOS 14.0, *) {
            let has_dict = Bundle.main.object(forInfoDictionaryKey: "NSLocationTemporaryUsageDescriptionDictionary") != nil
            if !has_dict {
                call.reject("Missing NSLocationTemporaryUsageDescriptionDictionary in Info.plist")
                return
            }
            guard let purpose = call.getString("purpose") else {
                call.reject("Missing purpose")
                return
            }

            location_manager.requestTemporaryFullAccuracyAuthorization(withPurposeKey: purpose) { error in
                if let error = error {
                    call.reject("Error when requesting temporary full location accuracy authorization: \(error)")
                } else {
                    let accuracy = (self.location_manager.accuracyAuthorization == .fullAccuracy) ? "full" : "reduced"
                    call.resolve(["value": accuracy])
                }
            }
        } else {
            call.reject("requestTemporaryFullAccuracyAuthorization is not available on iOS < 14")
        }
    }

    // -------------------------------------------------------------------------
    // CLLocationManagerDelegate
    // -------------------------------------------------------------------------

    /*
     * iOS 14+ combined delegate — handles both authorization and accuracy changes.
     * Fires report_change_authorization_status to resolve any pending auth calls.
     */
    @available(iOS 14.0, *)
    func locationManagerDidChangeAuthorization(_ manager: CLLocationManager) {
        report_change_authorization_status(manager.authorizationStatus)

        let accuracy = (manager.accuracyAuthorization == .fullAccuracy) ? "full" : "reduced"
        let changed: Bool
        if let current = current_location_accuracy_authorization {
            changed = (accuracy != current)
        } else {
            changed = true
        }
        current_location_accuracy_authorization = accuracy
        _ = changed
    }

    /*
     * Pre-iOS 14 delegate for authorization status changes.
     */
    func locationManager(_ manager: CLLocationManager, didChangeAuthorization status: CLAuthorizationStatus) {
        report_change_authorization_status(status)
    }
}